### PR TITLE
Enable full survey access

### PR DESCRIPTION
**Pull request by ddobie**

### CODE REVIEW

**Comment by ddobie** [comment]
@ajstewart this is a big one, but it needed to be because of all the changes in the full survey data specs.

Specifically,
1. We no longer use "epochs". The terminology is retained in all VAST software to save the hassle of changing everything, but they're more like groups than epochs. Specifically, each epoch no longer necessarily contains all fields, and each field can be observed multiple times in one epoch. The reason for this is that the pilot observations were scheduled such that VAST monopolised the telescope for as long as required to observe all fields, which is extremely inefficient and therefore not feasible for the full surveys. Instead, we're planning on redefining epochs to be all observations on a given day (or maybe week).
2. The field names no longer have an "A" on the end, which was a leftover from when observations were potentially going to be interleaved. Allowing access to data that uses both naming schemes is not trivial, but we can now handle it.

File: vasttools/query.py [hunk 1]
```diff
@@ -1893,17 +1919,16 @@ def _add_files(self, row: pd.Series) -> Tuple[str, str, str]:
...
+                                   )
             if conv_image_file.is_file():
                 image_file = conv_image_file
-                rms_file = Path(str(image_file).replace('.restored',
-                                                        '.restored.conv')
-                                                        )
-
+                rms_file = Path(str(rms_file).replace('.restored',
+                                                      '.restored.conv')
+                                )
```
Comment (ajstewart) [inline vasttools/query.py:1931]:
```suggestion
                rms_file = rms_file.with_suffix('.restored.conv')
```
Is `rms_file` already a Path here? If so you could do the suggestion.

File: vasttools/query.py [hunk 2]
```diff
@@ -2128,7 +2158,7 @@ def find_fields(self) -> None:
             )
         else:
             self.logger.info(
-                "%i/%i sources in VAST Pilot footprint.",
+                "%i/%i sources in VAST footprint.",
                 self.num_sources_searched,
                 prev_num
```
Comment (ajstewart) [inline vasttools/query.py:2163]: As changing anyway consider updating to modern string formatting method.

File: vasttools/query.py [hunk 3]
```diff
@@ -2173,84 +2203,113 @@ def _field_matching(
             Tuple containing the field information.
         """
 
+        self.logger.debug(row)
```
Comment (ajstewart) [inline vasttools/query.py:2206]: Log message has no context to help identify it.

File: vasttools/query.py [hunk 4]
```diff
@@ -2173,84 +2203,113 @@ def _field_matching(
...
-                )
-            else:
-                self.logger.info(
-                    "Source '%s' not in VAST Pilot footprint.",
-                    row['name']
-                )
-            return np.nan, np.nan, np.nan, np.nan, np.nan, np.nan
+            self.logger.info(
+                "Source '%s' not in the requested epoch footprint.",
+                row['name']
```
Comment (ajstewart) [inline vasttools/query.py:2222]:
```suggestion
                f"Source '{row['name']}' not in the requested epoch footprint."
```

File: vasttools/query.py [hunk 5]
```diff
@@ -2173,84 +2203,113 @@ def _field_matching(
...
+                the_fields = vast_fields
             else:
                 the_fields = fields
 
+            epoch_fields_names = self._epoch_fields.loc[i].index
+            stripped = False
+            if epoch_fields_names[0].endswith('A'):
+                self.logger.debug("Using stripped field names")
+                stripped = True
+                epoch_fields_names = epoch_fields_names.str.rstrip('A')
```
Comment (ajstewart) [inline vasttools/query.py:2251]: Can replace with the function you defined in survey was it? Might be useful in utils if it's also used here.

File: vasttools/query.py [hunk 6]
```diff
@@ -2260,6 +2319,9 @@ def _field_matching(
                        dateobs,
                        freqs
                        )
+        # If len(available_fields) == 0 for all epochs need to return nan
+        if len(epochs) == 0:
+            return [np.nan] * 7
```
Comment (ajstewart) [inline vasttools/query.py:2324]: Why is it 7 NaNs here and elsewhere?

File: vasttools/query.py [hunk 7]
```diff
@@ -2360,18 +2425,34 @@ def _build_catalog(self) -> pd.DataFrame:
...
+            self.logger.debug("Masking sources outside footprint")
+            mask = footprint_moc.contains(
                 self.coords.ra, self.coords.dec, keep_inside=False
             )
             if mask.any():
                 self.logger.warning(
-                    "Removing %i sources outside"
-                    " the VAST Pilot Footprint", sum(mask)
+                    "Removing %i sources outside the "
+                    "requested survey footprint", sum(mask)
```
Comment (ajstewart) [inline vasttools/query.py:2450]:
```suggestion
                    f"Removing {sum(mask)} sources outside the requested survey footprint"
```

File: vasttools/query.py [hunk 8]
```diff
@@ -2477,6 +2547,38 @@ def _get_epochs(self,
 
         return epochs
 
+    def _check_survey(self, epochs: list):
```
Comment (ajstewart) [inline vasttools/query.py:2550]:
```suggestion
    def _check_survey(self, epochs: list) -> None:
```

File: vasttools/query.py [hunk 9]
```diff
@@ -2477,6 +2547,38 @@ def _get_epochs(self,
...
+            self.vast_p1 = True
+        if len(epochs_set & set(P2_EPOCHS)) > 0:
+            self.vast_p2 = True
+        if len(epochs_set & set(full_epochs)) > 0:
+            self.vast_full = True
+
+        self.logger.debug(f"self.racs: {self.racs}")
+        self.logger.debug(f"self.vast_p1: {self.vast_p1}")
+        self.logger.debug(f"self.vast_p2: {self.vast_p2}")
+        self.logger.debug(f"self.vast_full: {self.vast_full}")
```
Comment (ajstewart) [inline vasttools/query.py:2580]: This function seems pretty important to test?

File: tests/test_query.py [hunk 10]
```diff
@@ -570,36 +570,66 @@ def test_init_failure_base_folder_not_found(self) -> None:
...
+            (True, False, True),
+            (True, True, False),
+            (False, True, False),
+        ]
+    )
+    def test_init_failure_stokes_v_tiles(self,
+                                         vast_pilot: bool,
+                                         vast_full: bool,
+                                         fails: bool,
+                                         mocker) -> None:
```
Comment (ajstewart) [inline tests/test_query.py:585]: I'm not suggesting to fix this everywhere, but the type hint for the mocker can be declared as so: https://pytest-mock.readthedocs.io/en/latest/remarks.html#type-annotations

File: vasttools/query.py [hunk 11]
```diff
@@ -2173,84 +2203,113 @@ def _field_matching(
...
 
             elif len(available_fields) == 1:
                 field = available_fields[0]
+                self.logger.debug("Selecting only available field")
 
             else:
                 field_indexes = [
                     field_centre_names[
-                        field_centre_names == f
+                        field_centre_names == f.rstrip('A')
```
Comment (ajstewart) [inline vasttools/query.py:2286]: The strip here, what's the reason that `available_fields` can't be sorted before this point?

**Review by ajstewart — CHANGES_REQUESTED** [review]
Some comments and suggestions on a first pass through. Don't envy dealing with that field name change!

File: vasttools/query.py (hunk 6 above)
Comment (ajstewart) [reply vasttools/query.py:2324]: Ah it's number of columns isn't it?

File: vasttools/query.py (hunk 1 above)
Comment (ddobie) [reply vasttools/query.py:1931]: The problem with that is that the `.restored.conv` is in the middle of the file name, not the suffix

Comment (ddobie) [reply vasttools/query.py:1931]: And as far as I can tell, there's no nice way to do a simple replace on a Path

File: vasttools/data/csvs/low_field_centres.csv [hunk 12]
```diff
@@ -1,4 +1,208 @@
 field,centre-ra,centre-dec
+VAST_0056-31A,14.12074696385262,-31.397409554915622
```
[the remark this answers is not in the document]
Comment (ddobie) [reply vasttools/data/csvs/low_field_centres.csv:2]: It's easier to remove the A when it's not needed than it is to add the A when it is needed

File: vasttools/query.py (hunk 11 above)
Comment (ddobie) [reply vasttools/query.py:2285]: I suspect this strip might actually be superfluous now that I'm looking at it closer. `available_fields` is built from `the_fields` which has already been stripped.

File: vasttools/query.py (hunk 1 above)
Comment (ajstewart) [reply vasttools/query.py:1931]: Ah of course!

File: vasttools/query.py (hunk 11 above)
Comment (ddobie) [reply vasttools/query.py:2284]: Available fields isn't always stripped, but is matched against a list of stripped fields. Stripping as needed vs doing it all at once should be equally fast.

**Review by ajstewart — APPROVED** [review]
Thanks for adding the tests.

If you're satisfied it's all working then it looks good to me!

**Pull request merged**